### PR TITLE
Fix numpy reader returning unread sample data

### PR DESCRIPTION
**Pull request by jantonguirao**
## Category:
**Bug fix** (*non-breaking change which fixes an issue*)

## Description:
`fn.readers.numpy` could hand out sample buffers that were never read. No error was raised: the
operator reported the correct shape and dtype, and the read was reported as full-length, but the
contents were whatever happened to be in the freshly allocated buffer - zeros, or stale heap data.

Two places decide whether the sample data still needs reading, and they disagreed.

`NumpyLoader::ReadSample` defers the read based on the actual stream:

```c++
if (!opts.use_mmap || !current_file->CanMemoryMap()) {
  target.current_file = std::move(current_file);   // data NOT read yet; read it later
} else {
  auto p = current_file->Get(nbytes);              // mmap path; data available now
  ...
}
```

`NumpyReaderCPU::Prefetch`, which performs that deferred read, gated on the user-facing argument
instead:

```c++
if (!dont_use_mmap_)
  return;                                          // skips the deferred read entirely
```

So when the stream was not memory mappable but `dont_use_mmap` was not requested, the loader
deferred the read and the operator never performed it.

This affects every `s3://` path, because `S3FileStream` does not override `CanMemoryMap()` and so is
never mappable, and local files whenever the mmap reservation cannot be satisfied - `copy_read_data_`
is `dont_use_mmap_ || !mmap_reserver_.CanShareMappedData()`.

The fix drops the early return. `use_o_direct` is unaffected; it already requires `dont_use_mmap`.

Once the loop runs for every batch, stopping at the first sample with no `current_file` is no longer
correct, so that `break` becomes a `continue`: a batch can mix samples the loader already provided
(memory mapped, or skipped by the image cache) with samples that still need a deferred read -
reachable today with `files=` listing both local and `s3://` paths - and the later ones would be
handed out unread. The other stop condition, on consecutive entries pointing at the same object,
stays a `break`: those are the `pad_last_batch` duplicates and they always form a contiguous tail.

### Affected modules and functionalities:
- `dali/operators/reader/numpy_reader_op.cc`: `NumpyReaderCPU::Prefetch` no longer returns early and
  skips already-provided samples individually. Also drops `previous_path`, a local in the same loop
  that has been unused since it was introduced.
- `dali/test/python/reader/test_numpy.py`: new regression test.

### Key points relevant for the review:
- The regression test has to reach the deferred path with `dont_use_mmap` unset, which locally means
  making `FileStream::MappingReserver` fail. It drains the process-wide reservation pool with readers
  that are only built and never run - the reservation is taken in the loader constructor, so this
  costs no reads, no file descriptors and no mappings - and then mops up the remaining entries one at
  a time. The reservation is granted all or nothing, so the test does not assume it worked: after
  each run it checks `/proc/self/maps` to see whether that reader actually memory mapped its input,
  and stops at the first one that did not. That reader is the one the assertion runs against.
- Verified against a build without the fix: the test fails with an all-zero sample on a clean pool,
  with 1/3/20/1024 pool entries already held by another live reader, and with the pool already fully
  exhausted by an outsider. The same drain with `dont_use_mmap=True` on the reader under test - a
  configuration the unfixed `Prefetch` already handles - returns the correct data, which is what the
  test will see once the fix is in.
- Cost and blast radius, measured: ~0.06 s, open file descriptors unchanged before/after and +3 at
  peak, and the whole reservation pool handed back in the `finally` (`MappingReserver` is an
  in-process counter freed by its destructor, so dropping the pipeline references is enough - no
  mapping is ever created and no `gc.collect()` is needed). Running the whole `reader/test_numpy.py`
  module in one interpreter gives the same error count with and without the new test.

### Tests:
- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

`test_numpy.py: test_sample_data_is_read_when_mmap_is_unavailable`

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-4887

### CODE REVIEW

**Comment by jantonguirao** [comment]
!build

File: dali/test/python/reader/test_numpy.py [hunk 1]
```diff
@@ -1220,3 +1220,38 @@ def test_shuffle_after_epoch_seed_numpy_default_reproducible():
...
         ), "Without explicit seed, default behavior should be reproducible (backward compat)"
+
+
+def test_sample_data_is_read_when_mmap_is_unavailable():
+    """Samples must be read even when the loader falls back to a non-mapped stream.
+
+    NumpyLoader::ReadSample defers the read whenever the stream cannot be memory mapped, which
+    happens for remote storage and, locally, when the mmap reservation cannot be satisfied.
+    NumpyReaderCPU::Prefetch used to skip that deferred read unless dont_use_mmap was requested,
+    so the sample was handed out as a never-written buffer. See DALI-4887.
```
Comment (JanuszL) [inline dali/test/python/reader/test_numpy.py]:
```suggestion
    so the sample was handed out as a never-written buffer..
```

**Review by JanuszL — APPROVED** [review]

File: dali/test/python/reader/test_numpy.py [hunk 2]
```diff
@@ -1220,3 +1221,50 @@ def test_shuffle_after_epoch_seed_numpy_default_reproducible():
...
+        try:
+            pipe = Pipeline(batch_size=1, num_threads=1, device_id=None)
+            with pipe:
+                pipe.set_outputs(fn.readers.numpy(file_root=test_data_root))
+            pipe.build()
+            assert_array_equal(to_array(pipe.run()[0][0]), ref)
+        finally:
+            # release the reservation so the rest of the suite keeps using mmap
+            del hog
+            gc.collect()
```
Comment (rostan-t) [inline dali/test/python/reader/test_numpy.py]: Why is it needed to force `gc.collect()` here? Does `munmap` get deferred, preventing the rest of the suite to use mmap?

File: dali/operators/reader/numpy_reader_op.cc [hunk 3]
```diff
@@ -263,8 +263,12 @@ void NumpyReaderCPU::Prefetch() {
                       DomainTimeRange::kRed);
   DataReader<CPUBackend, NumpyFileWrapper, NumpyFileWrapper, true>::Prefetch();
 
-  if (!dont_use_mmap_)
-    return;
+  // Read the samples that the loader could not provide by memory mapping. Note that this is not
+  // the same as `dont_use_mmap_`: NumpyLoader::ReadSample decides based on the actual stream
+  // (`!opts.use_mmap || !current_file->CanMemoryMap()`), so a sample can need a deferred read even
+  // when the user did not ask for `dont_use_mmap`, e.g. for remote storage, which is never
+  // mappable, or when the mmap reservation could not be satisfied. Samples that the loader already
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/operators/reader/numpy_reader_op.cc:270]: Fixed — that `break` is a `continue` now, so a batch that mixes samples the loader already provided (memory mapped ones, image-cache skips) with deferred ones reads every deferred sample instead of stopping at the first provided one.

The other `break`, on `curr_batch[idx - 1] == curr_batch[idx]`, stays a `break` on purpose: those are the `pad_last_batch` duplicates and they always form a contiguous tail — `Loader::ReadOne` only starts padding once the shard is exhausted, and the padding branch returns before touching `shards_`, so it cannot un-exhaust mid-batch. Turning that one into a `continue` would re-read an already consumed stream and reintroduce the race fixed in #4848. I added a note to that effect next to it.

File: dali/test/python/reader/test_numpy.py [hunk 4]
```diff
@@ -1220,3 +1220,38 @@ def test_shuffle_after_epoch_seed_numpy_default_reproducible():
...
+    try:
+        with open("/proc/sys/vm/max_map_count") as f:
+            max_map_count = int(f.read())
+    except OSError:
+        raise SkipTest("/proc/sys/vm/max_map_count is not readable")
+
+    # FileLoader reserves initial_buffer_fill_ mappings up front and falls back to copying reads
+    # when that fails; the limit is max_map_count / 2 (mmaped_file.cc). initial_fill only feeds
+    # initial_buffer_fill_ when random_shuffle is on.
+    initial_fill = max_map_count // 2 + 1
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/test/python/reader/test_numpy.py]: You were right, and the consequence was worse than expense: that version deterministically failed its own reservation by asking for `pool + 1`, so the initial buffer fill opened ~32k `StdFileStream` handles and the CI job died with EMFILE, taking 628 unrelated tests down with it.

The number of live samples no longer scales with the host limit. The pool is claimed by readers that are only built and never run — the reservation is taken in the loader constructor, so it costs no reads, no file descriptors and no mappings — and the reader under test keeps `initial_buffer_fill_` at 1. Measured on the current version: open file descriptors unchanged before/after and +3 at peak, ~0.06 s, and running the whole `reader/test_numpy.py` module in one interpreter gives the same error count with and without the new test.

File: dali/test/python/reader/test_numpy.py [hunk 5]
```diff
@@ -1220,3 +1220,38 @@ def test_shuffle_after_epoch_seed_numpy_default_reproducible():
...
+        # a single sample, so random_shuffle cannot change what we get back
+        np.save(os.path.join(test_data_root, "sample.npy"), ref)
+
+        pipe = Pipeline(batch_size=1, num_threads=1, device_id=None)
+        with pipe:
+            pipe.set_outputs(
+                fn.readers.numpy(
+                    file_root=test_data_root, random_shuffle=True, initial_fill=initial_fill
+                )
+            )
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/test/python/reader/test_numpy.py]: Done — the test uses `@pipeline_def` now.

File: dali/test/python/reader/test_numpy.py (hunk 1 above)
Comment (jantonguirao) [reply dali/test/python/reader/test_numpy.py]: Applied — the tracker reference is gone from the docstring.

File: dali/test/python/reader/test_numpy.py [hunk 6]
```diff
@@ -1220,3 +1221,50 @@ def test_shuffle_after_epoch_seed_numpy_default_reproducible():
...
+        # reader below is then forced onto the non-mapped path with a single sample in flight.
+        hog = Pipeline(batch_size=1, num_threads=1, device_id=None)
+        with hog:
+            hog.set_outputs(
+                fn.readers.numpy(
+                    file_root=test_data_root, random_shuffle=True, initial_fill=whole_pool
+                )
+            )
+        hog.build()
+        try:
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/test/python/reader/test_numpy.py]: Confirmed, and I could reproduce it: with one other live reader holding ten pool entries, the all-or-nothing request for the whole pool returned 0, the reader under test memory mapped normally, and the test passed green against a build that still has the bug.

It no longer assumes the reservation, it observes the result. The pool is drained with a halving ladder of never-run readers, then the remaining entries are mopped up one at a time, checking `/proc/self/maps` after each run to see whether that reader actually memory mapped its input. A reader that did holds its entry for as long as it is alive, so the loop makes progress; it stops at the first reader the exhausted pool pushed onto the non-mapped path, and that is the one the assertion runs against.

Against a build without the fix the test now fails on a clean pool, with 1/3/20/1024 entries pre-held by another live reader, and with the pool already fully exhausted by an outsider — the previous version passed vacuously in several of those. The batch is 4 samples with distinct contents now, so a partially read batch is caught as well.

File: dali/test/python/reader/test_numpy.py (hunk 2 above)
Comment (jantonguirao) [reply dali/test/python/reader/test_numpy.py]: No — nothing is deferred and no `munmap` is involved. It was defensive, and it is gone.

The readers that claim the pool are never run, so they never map anything. `FileStream::MappingReserver` only bumps a process-global counter (`dali_reserved_mv_cnt` in `dali/util/mmaped_file.cc`) and its destructor hands the entries back. The chain is `Pipeline.__del__` → `_shutdown()` → `self._pipe = None` → C++ pipeline → loader → `~MappingReserver`, all synchronous. `Pipeline` is not in a reference cycle (`Pipeline._pipes` keeps only a `weakref`), so plain refcounting finishes the job: with the cyclic collector disabled the pool is free again on the very next statement. The new version just drops the references in the `finally`.

File: dali/test/python/reader/test_numpy.py [hunk 7]
```diff
@@ -1220,3 +1220,88 @@ def test_shuffle_after_epoch_seed_numpy_default_reproducible():
...
+                    break
+            else:
+                raise AssertionError("could not exhaust the mmap reservation pool")
+
+            for idx, ref in enumerate(refs):
+                assert_array_equal(to_array(out[idx]), ref)
+        finally:
+            # MappingReserver is an in-process counter released by its destructor, so dropping the
+            # last reference to each pipeline hands the entries back. No mapping is involved.
+            keep_alive.clear()
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/test/python/reader/test_numpy.py:1307]: This isn't a new operator or a new operator argument. `fn.readers.numpy` itself is unchanged, and its CPU/GPU variants, batch sizes, edge cases and error paths are already covered throughout this file. This is a targeted regression test for one internal path in `NumpyReaderCPU::Prefetch` — the deferred read that the operator used to perform only when `dont_use_mmap` was set.

On the specific items:

- **GPU variant**: the GPU reader has no such path. `NumpyLoaderGPU` never memory maps (`opts.use_mmap = false` in `NumpyFileWrapperGPU::Reopen`) and `NumpyReaderGPU::Prefetch` has no `dont_use_mmap_` short-circuit, so a GPU version of this test would assert nothing about the bug.
- **Invalid-input rejection**: there is no input to reject here. The bug was a silent success — correct shape, correct dtype, full-length read reported, uninitialised contents.
- **`batch_size=1`**: the batch is deliberately 4 samples rather than 1, so that a batch which is only partially read is caught. That is precisely the property the `break` → `continue` change turns on, and `batch_size=1` cannot express it.

**Comment by jantonguirao** [comment]
!build

**Review by rostan-t — APPROVED** [review]
